public/talos/v1.12/configure-your-talos-cluster/system-configuration/oom.mdx

```diff
@@ -0,0 +1,102 @@
+---
+title: "OOM handler"
+description: "Configuring userspace out-of-memory handler."
+---
+
+Talos Linux includes a configurable userspace low-memory monitor supplementing Linux kernel built-in OOM killer.
+This controller enables early detection of heavy memory use and helps prevent machine lock-up due to out-of-memory,
+which is especially important to enhance the stability of some special cases making control plane more prone to OOM,
+such as single-node clusters or scheduling pods on control plane nodes.
+
+While the kernel is already capable of resolving OOMs when they happen, this comes at a cost of machine
+taking significant time before returning to being responsive, because the kernel OOM killer only kicks in
+when there are no free pages to allocate for a process. Userspace services are meant to detect
+and prevent upcoming OOM by continuously monitoring available resources as well as PSI - a relative measure
+of the demand on various system resources - to detect when the system starts experiencing lack of available memory.
```

Comment:
I'd rephrase this for clarity – Maybe something like
```suggestion
While the Linux kernel is already capable of handling low-memory situations, the kernel OOM killer
only kicks in when the kernel has completely run out of free pages to allocate for a process – at
which point a machine is already struggling (or unresponsive), and will take a while to recover.

Instead of letting the machine run out of memory/start to struggle, userspace OOM handlers can
continuously monitor available resources, and proactively prevent OOM situations before the machine
starts to struggle.
```